lib/chai/core/assertions.js

```diff
@@ -2447,130 +2447,214 @@ module.exports = function (chai, _) {
   Assertion.addMethod('keys', assertKeys);
   Assertion.addMethod('key', assertKeys);
 
+  // TODO: Document and move to utilities
+  function createErrCriteria(errLike, errMsgMatcher, negate) {
+    var cri = {
+      errLike: errLike,
+      errLikeType: _.type(errLike).toLowerCase(),
+      errMsgMatcher: errMsgMatcher,
+      errMsgMatcherType: _.type(errMsgMatcher).toLowerCase(),
+    };
+
+    // Short-circuit if no criteria is defined.
+    if (cri.errLikeType === 'undefined' &&
+        cri.errMsgMatcherType === 'undefined') {
+      return cri;
+    }
+
+    if (negate) {
+      throw TypeError('errLike and errMsgMatcher must both be undefined when' +
+                      ' negate is true');
+    }
+
+    if (cri.errMsgMatcherType === 'undefined') {
+      // Handle `errMsgMatcher` being passed as the first argument.
+      if (cri.errLikeType === 'string' || cri.errLikeType === 'regexp') {
+        cri.errMsgMatcher = cri.errLike;
+        cri.errMsgMatcherType = cri.errLikeType;
+        cri.errLike = undefined;
+        cri.errLikeType = 'undefined';
+      }
+      return cri;
+    }
+
+    if (cri.errLikeType !== 'function') {
+      throw TypeError('errLike must be a function when errMsgMatcher is' +
+                      ' defined');
+    }
+
+    if (cri.errMsgMatcherType !== 'string' &&
+        cri.errMsgMatcherType !== 'regexp') {
+      throw TypeError('errMsgMatcher must be a string or regexp when' +
+                      ' defined');
+    }
+
+    return cri;
+  }
```

Comment:
I'd be interested to see these functions move into https://github.com/chaijs/check-error where they can be tested for their surface area, and can be reviewed+merged+shipped easier there, thereby making this PR smaller.

Reply:
I thought about that. `createErrCriteria` could probably be moved to https://github.com/chaijs/check-error, but it'd be the only function there, without any real non-Chai applicability. `assertErrCriteria` is using a bunch of other Chai assertions internally, so I don't think it can be cleanly moved to a separate module until v5 when assertions are better modularized.